### PR TITLE
Added backend support to get locations by search

### PR DESCRIPTION
**Pull request by ThomasPFMuller**
This PR adds the backend support for interacting with the geocode API (based on OSM). The methods have been tested by hand but creating automated tests for the methods is finnicky and I have already spent too much time on it. I will create automated tests in a later sprint.

The added methods store responses from API as instances of the Location class.

### CODE REVIEW

File: app/src/main/java/com/github/wanderwise_inc/app/network/DirectionsApiService.kt [hunk 1]
```diff
@@ -46,8 +46,8 @@ data class DirectionsResponseBody(val routes: List<Route>) {
   }
 }
 
-/** Factory for creating `ApiService` design patterns uwu */
-object ApiServiceFactory {
+/** Factory for creating `ApiService` design patterns for Directions API uwu */
```
Comment (ethangraham2001) [inline app/src/main/java/com/github/wanderwise_inc/app/network/DirectionsApiService.kt:49]: thanks for leaving in my uwu

File: app/src/main/java/com/github/wanderwise_inc/app/network/LocationsApiService.kt [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  fun createLocationsApiService(): LocationsApiService {
+    val retrofit =
+        Retrofit.Builder()
+            .baseUrl(BASE_URL)
+            .addConverterFactory(GsonConverterFactory.create())
+            .build()
+
+    return retrofit.create(LocationsApiService::class.java)
+  }
+}
```
Comment (ethangraham2001) [inline app/src/main/java/com/github/wanderwise_inc/app/network/LocationsApiService.kt:44]: Haven't played around with retrofit in a little while, but this is looking great to me.

File: app/src/main/java/com/github/wanderwise_inc/app/viewmodel/ItineraryViewModel.kt [hunk 3]
```diff
@@ -146,8 +148,25 @@ open class ItineraryViewModel(
...
+    }
+  }
+
+  fun getPlacesLiveData(): LiveData<List<Location>> {
+    return locationsLiveData
+  }
+
   /** @brief get a Flow of the user location updated every second */
-  fun getUserLocation(): Flow<Location> {
+  fun getUserLocation(): Flow<android.location.Location> {
```
Comment (ethangraham2001) [inline app/src/main/java/com/github/wanderwise_inc/app/viewmodel/ItineraryViewModel.kt:169]: Thanks for changing this - I've been getting confused by the android class sharing the same name

File: app/src/main/java/com/github/wanderwise_inc/app/data/LocationsRepositoryImpl.kt [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+              }
+
+              override fun onFailure(call: Call<List<Place>>, t: Throwable) {
+                Log.d(DEBUG_TAG, "request failed! ${t.message}")
+                resultLiveData.value = null // or any other value that represents a network error
+              }
+            })
+
+    return resultLiveData
+  }
```
Comment (ethangraham2001) [inline app/src/main/java/com/github/wanderwise_inc/app/data/LocationsRepositoryImpl.kt:49]: Looks correct to me

File: app/src/test/java/com/github/wanderwise_inc/app/viewmodel/ItineraryViewModelTest.kt [hunk 5]
```diff
@@ -224,9 +230,50 @@ class ItineraryViewModelTest {
...
+
+    advanceUntilIdle()
+
+    val placesLocations = itineraryViewModel.getPlacesLiveData().value
+    assertNotNull(placesLocations)
+  }
+
+  @ExperimentalCoroutinesApi
+  @Test
+  fun getPlacesLiveData() = runTest {
```
Comment (ethangraham2001) [inline app/src/test/java/com/github/wanderwise_inc/app/viewmodel/ItineraryViewModelTest.kt:250]: high quality test

**Review by ethangraham2001 — APPROVED** [review]
Really good work. One question perhaps - you're not calling this from any UI right now. Do you think you could clarify how you tested/how you intend to test this in practice?

**Comment by ThomasPFMuller** [comment]
Yes of course, so the API call works by sending HTTP GET requests with retrofit, as you surely know. I can build the same request and enter it in my browser to see the JSON the server has responded with. I can then manually check if the response is correctly deserialized to a List<Location> as intended. I also tested it by creating a composable with a search bar and a google map in which I could enter a name, and the composable moved the camera to the corresponding latitude and longitude.

**Comment by ThomasPFMuller** [comment]
closes #258 and #257

**Pull request merged**